### PR TITLE
Use right handed coordinate system for rotation matrices and quaternions. Also fixed Euler angles (XYZ convention).

### LINKED ISSUE #6816: Rotation matrix multiplication order should be ZYX
i tried godot to verify some 3d transformations and found out that they were 'incorrect'.
it seems like godot uses XYZ order for rot. matrix multiplication but the correct order is ZYX (like in blender, and glm's quaternions)

im not an expert in 3d math but i read articles in internet that the correct order for euler angles is ZYX (like in quaternions) but many apps use XYZ (incorrect)

**Steps to reproduce:**
1. create an entity in blender, translate by 2,3,4 then rotate by 45,45,45 deg 
   compare the position with godot, try switch rotation order in blender (XYZ, YXZ - will give different position)
2. use this code with glm

```

#include <iostream>
#include <glm/vec3.hpp> // glm::vec3
#include <glm/vec4.hpp> // glm::vec4
#include <glm/mat4x4.hpp> // glm::mat4
#include <glm/gtc/matrix_transform.hpp> // glm::translate, glm::rotate, glm::scale, glm::perspective
#include <glm/gtc/matrix_inverse.hpp>
#include <glm/gtc/constants.hpp> // glm::pi
#include <glm/gtc/quaternion.hpp>
#include <glm/ext.hpp>
#include <glm/gtx/string_cast.hpp>
#include <glm/gtx/matrix_decompose.hpp>
#include <glm/gtx/quaternion.hpp>

using namespace std;
using namespace glm;

int main()
{
         mat4 rm = mat4(); // identity matrix
    // ZYX order (yaw, pitch, roll); "XYZ" in blender.
    rm = rotate(rm, radians(45.0f), vec3(0.0, 0.0, 1.0)); /// change values here to have "XYZ" order (wrong)
    rm = rotate(rm, radians(45.0f), vec3(0.0, 1.0, 0.0));
    rm = rotate(rm, radians(45.0f), vec3(1.0, 0.0, 0.0));
    //mat4 tm = mat4();
    mat4 tm = translate(vec3(2., 3., 4.));
    mat4 sm = mat4();

    mat4 modelMatrix = tm * rm * sm;
        mat4 m = modelMatrix;

    cout << "model matrix:" << endl;
    cout << to_string(m) << endl << endl;



        // Verifing it with quaternion rotation

    // quaternion rm test

    vec3 EulerAngles = radians(vec3(45, 45, 45)); // radians
    quat q1 = quat(EulerAngles);

    rm = mat4();
    rm = toMat4(q1); // quat to Rm
    m = tm * rm * sm;

    cout << "model matrix from quat:" << endl;
    cout << to_string(m) << endl << endl;

    cout << "quat:" << endl;
    cout << to_string(q1) << endl << endl;

        return 0;

}

```

hope it compiles

**Link to minimal example project** (optional but very welcome):

### PR DESCRIPTION
**Pull request by ghost**
Clarified the Euler angle convention with comments, and added references. Also made two cosmetic changes in quad.cpp.

Fixes #6816

### CODE REVIEW

**Comment by tagcup** [comment]
PTAL.

**Comment by tagcup** [comment]
There are a few more things that I'd like to do, but I'm not sure I should open another PR for those or not. Namely:

* Fix the sign of the angle for all users of `rotate` and `rotated` and remove the workaround. It turned out that there aren't that too many users of these functions, so hopefully I can manage to do this change without breaking stuff :)
* Remove `rotate`, `rotated`, `scale`, `scaled` functions from `Matrix3` as these functions make sense only for `Transform` and not for a general `Matrix3` (proper matrix rotation is `R.M.Rᵀ`, not `M.R` which just adds another rotation to object's `Transform` instead of rotating the matrix itself).
* For rotating basis vectors in a `Transform`, make use of `Translate::rotate_basis` instead. I'm not exactly sure how this is used (I'm not familiar with rendering stuff, I'll check later on) but I believe this is a passive rotation, so I'll fix the angle in its definition.

All these changes will effect 3D math. There are similar issues with 2D, but I'll do that in a different PR. In any case, I'll proceed to do more stuff if this PR gets accepted.

**Comment by tagcup** [comment]
The changes in this PR are breaking, so I hope someone from the Godot team can comment on. Depending on the input, I'll continue implementing what is left, and then move on to 2D.

These changes are required for mathematical correctness, and to align Godot functions that generate rotations with Unity and other engines.

**Comment by reduz** [comment]
We can try to merge for 3.0 at compatibility breaking time. We need to
note, however, what has changed

**Comment by tagcup** [comment]
Thanks! I'll go ahead and implement what is left for 3D, and post a mini-changelog here at the end.

Since we can break compatibility at this point, would you like to change the convention for Euler angles too? With the current changes in the PR, Godot will be using XYZ convention properly. Much like Blender, which defaults to XYZ and uses a right-handed coordinate system.

Unity uses YXZ convention, and while the documentation is not clear, it appears that Unreal is using YXZ too (the order the angles are written in [FRotator doc](https://docs.unrealengine.com/latest/INT/API/Runtime/Core/Math/FRotator/index.html) suggests that). However, they also use a left-handed coordinate system (which is a shocker to me), so I don't really think Godot has to follow them.

**Comment by tagcup** [comment]
I think I'm done for this PR involving 3D, so please take another look.

There are a few more things that can be done as I listed above, but I think they're more suited for a different PR as they're minor stuff like improvement of the API or documentation. I will submit a different PR for those in the future.

**Comment by tagcup** [comment]
Can you hold on? I was double-checking everything by manually constructing the rotation matrices in Mathematica, and I realized a problem. The current implementation in the PR is actually using ZYX (X-first, Z-last) thanks to the [error](https://en.wikipedia.org/wiki/Euler_angles#Rotation_matrix) in wikipedia! Whoever wrote that section thinks that a rotation matrix can be decomposed as three rotation matrices `R = X(α) Y(β) Z(γ)`, with `X(α)` being the first rotation! The confusion flows to the tables for rotation matrices below which is self-conflicting: the subscript is the correct sequence, the ordering of the matrices is wrong.

Let me revise this PR such that we use XYZ (Z-first), which is the Blender-default. I will also remove references to Wikipedia in the comments.

**Comment by tagcup** [comment]
The code turned out to be OK (apparently, I did notice the ordering was "wrong" in the first place).

We are essentially using the X₁ Y₂ Z₃ from [this table](https://en.wikipedia.org/wiki/Euler_angles#Rotation_matrix) , and it does correspond to the rotation matrix `R = X(θ₁).Y(θ₂).Z(θ₃)`. where θ₃ is the angle of the *first* rotation. This is the verification:

mathematica



The [NASA doc](https://ntrs.nasa.gov/archive/nasa/casi.ntrs.nasa.gov/19770024290.pdf) also uses the same convention (the matrix on page A-2 is identical to the one in the screenshot); they construct the matrix in pages 2-4. Essentially, they're defining the rotation matrix through x = Mx' (Eq. 1 on page 2, so we're rotating x' into x) which means we're first rotating around Z, then Y, then X (see Eq. 6). But the angles are numbered as if we're rotating x to x' (Eq. 2 is the X rotation, which is the last one in M = X Y Z). Now I'm not sure why they would do [that](https://en.wikipedia.org/wiki/Rotation_matrix#General_rotations), it's probably something related to engineering applications, but the mathematics is clear as long as we define the Euler angles explicitly by writing down the full equation like "`v_rotated = R v_original` where `R = X(ax).Y(ay).Z(az)`" which is what I did. And only after getting the angles right, I renamed the variables such that our naming of the angles matches the naming in the references.

It's really easy to make a mistake here, so I kept the references, and kept this convention (θ₃ is the first angle) such that it is easy to check the implementation against the references. I also added a clarification about this.

So in summary, things checked out OK.

**Comment by bojidar-bg** [comment]
Assinging reduz for review, since that's pretty much core stuff :smile:

**Comment by tagcup** [comment]
Thanks!
Once/if this lands, I'll implement other stuff I listed in my [2nd comment in this issue](https://github.com/godotengine/godot/pull/6865#issuecomment-254934040), hopefully in time for 3.0.

**Comment by tagcup** [comment]
Just as a note for other devs, I noticed that C# bindings (and maybe others too, if any) have been translating the current (problematic) code: [Quat.cs](https://github.com/neikeq/GodotSharp/blob/master/GodotSharp/Core/Quat.cs), for example.

**Comment by tagcup** [comment]
Congrats on merging the gles3 branch! I rebased my changes on the latest master.

**Comment by reduz** [comment]
Merged! feel free to do the same work in 2D!
